src/psyclone/psyir/tools/call_tree_utils.py

```diff
@@ -0,0 +1,415 @@
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+''' This module provides tools analyse the sequence of calls
+across different subroutines and modules.'''
+
+from psyclone.core import Signature, VariablesAccessInfo
+from psyclone.parse import ModuleManager
+from psyclone.psyGen import BuiltIn, Kern
+from psyclone.psyir.nodes import (Call, Container, IntrinsicCall, Reference)
+from psyclone.psyir.symbols import (ArgumentInterface, DefaultModuleInterface,
+                                    ImportInterface)
+from psyclone.psyir.tools.read_write_info import ReadWriteInfo
+
+
+# pylint: disable=too-few-public-methods
+class CallTreeUtils():
+    '''This class provides functions functions to analyse the sequence of
+    calls.
+    '''
+
+    # ------------------------------------------------------------------------
+    def _compute_all_non_locals(self, routine):
+        # pylint: disable=too-many-branches
+        '''This function computes all non-local access of the specified
+        routine node. It returns a list of 3-tuples containing:
+        - the type of the non-local as string, one of:
+          - routine: A routine that is called
+          - unknown: an imported symbol, which could still be a function or
+                     an array. This will be resolved later when checking
+                     the module the symbol is imported from.
+          - reference: a variable access
+        - the name of the module the symbol is imported from
+        - the signature of the access
+
+        :param routine: the routine for which to collect all non-local
+            accesses
+        :type routine: :py:class:`psyclone.psyir.nodes.Routine`
+
+        :returns: list of non-local references
+        :rtype: List[Tuple[str, str, :py:class:`psyclone.core.Signature`]]
+
+
+        '''
+        non_locals = []
+
+        outer_module = routine.ancestor(Container)
+
+        for access in routine.walk((Kern, Call, Reference)):
+            # Builtins are certainly not externals, so ignore them.
+            if isinstance(access, BuiltIn):
+                continue
+
+            if isinstance(access, Kern):
+                # A kernel is a subroutine call from a module:
```

Comment:
Please add a TODO here for #2054 as hopefully that will remove the need for this branch (it will be a `Call`).

Reply:
Done.